skycoin/skywire
# Fix updater [WIP]

Did you run `make format && make check`?
yes

The initial work adds a much abbreviated update process in place of the existing one, without changing much.

Needed / Proposed Changes:

- Remove version checks completely
- prompt if repo configuration is needed
- otherwise, update without prompting or checking the version, via `apt`

#### Version Checks

`apt update && apt install skywire-bin` will update skywire to the latest version or install it if it is not installed. Packages are not re-installed, by default, with `apt install` (correct me if I am wrong).

The version checking is entirely unnecessary, all of it. Yet they occupy a substantial part of the code that I was not able to remove it without breaking things.

#### Prompt for repository configuration

If the repository used for updating the .deb packages is not already configured in the software sources, it is highly abnormal to preform this configuration without the explicit consent of the user.

Secondly, the way this configuration was previously done was no longer congruent with the skybian images (on my fork / in my open PR), and actually conflicting with that (wouldn't detect the existing config and then would redundantly set it in a different file).

This would be the open question - the repository configuration in my skybian fork is provided by the skybian package, which is architecture-specific to arm architectures. This repository configuration could be made a seprate .deb package (and could include the repository signing key) and would then be able to work on amd64. There is no real reason not to do this, from my perspective, and the advantages for testing should be obvious.

So there needs to be a prompt to preform this configuration ; a .deb package would be downloaded from the archive of the apt repo server, and then upon it's installation, skywire-bin would be available to install via `apt`

#### otherwise update without prompting

Clicking the update button should either not prompt the user for confirmation if the repository is set, or it should  only give a generic "are you sure you want to install / update?" confirmation dialog.

The visor should restart at the end of the update

#### Testing

requires:

* visor is run as root
* .deb based distro

```
git clone https://github.com/the-skycoin-project/skywire
git checkout fix-updater
make run-source-sudo
```

* open [http://127.0.0.1:8000](http://127.0.0.1:8000)
* go to the visor page
* click update from the drop-down menu
* ui prompt for adding the repository configuration (installing the package described earlier)
* click yes, software sources are configured as they would be in skybian
* skywire is installed

So in theory, when this can take place, the update button _will also function as the installer_ or a wrapper for the process which installs it, and _that should be the easiest thing to test_

#### Still needed

In order for the above outlines process to be robust in terms of using the key that the visor was already running with, when it is being installed by the updater button the key should be passed via environmental variable to skkywire-autoconfig, which will further pass it to the config gen command.

Additionally, the autoconfig script assumes skywire was already running as a systemd service, so there should be some thought given to the transition between the visor running from the source code to the visor running from the installation.
